### PR TITLE
Marketplace - Only use CloudSQL when connection is provided

### PR DESCRIPTION
**Pull request by Ark-kun**
Fixes https://github.com/kubeflow/pipelines/issues/3798

This PR only fixes the CloudSQL issue. The GCS bucket issue is fixed in other PRs.

The new logic is: Since CloudSQL cannot function without `cloudsqlInstanceConnectionName`, the presence of this value now controls whether CloudSQL replaces the in-cluster MySQL.

### CODE REVIEW

**Comment by rmgogogo** [comment]
this may can't solve things but introduce other issue. You may check the enabled usage.

GCS bucket also uses it.

**Comment by rmgogogo** [comment]
Fixes #3797 

Expect this one to fix 3797 together

**Comment by rmgogogo** [comment]
https://github.com/kubeflow/pipelines/search?q=Values.managedstorage.enabled&unscoped_q=Values.managedstorage.enabled

We have more place to be solved. I'm wondering whether it's a good fix.

**Comment by Ark-kun** [comment]
> Fixes #3797
> Expect this one to fix 3797 together

I'm not sure I understand how this PR fixes #3797 (wrong value in the `workflow-controller-configmap`).

**Comment by Ark-kun** [comment]
This PR only fixes the CloudSQL issue. The GCS bucket issue is fixed in other PRs.

The new logic is: Since CloudSQL cannot function without `cloudsqlInstanceConnectionName`, the presence of this value now controls whether CloudSQL replaces the in-cluster MySQL.

**Comment by Bobgy** [comment]
I think we should discuss both this and https://github.com/kubeflow/pipelines/pull/3803 together.

One reason we explicitly kept sql connection name and gcs bucket name coupled was because artifact data in the db are links to locations in the gcs bucket. Therefore, after user first specifies a db and a bucket.
/cc @IronPan 

We'd want them to always use the same combination, otherwise, they will find the data in the db doesn't match those in the bucket.

If we now consider CloudSQL as optional (I guess users may want this too, because gcs obviously are used more often). Maybe we should just let users choose both names independently.

After all, cluster manager should remember which cloudsql and which gcs bucket were used together.

**Comment by rmgogogo** [comment]
> I think we should discuss both this and #3803 together.
> 
> One reason we explicitly kept sql connection name and gcs bucket name coupled was because artifact data in the db are links to locations in the gcs bucket. Therefore, after user first specifies a db and a bucket.
> /cc @IronPan
> 
> We'd want them to always use the same combination, otherwise, they will find the data in the db doesn't match those in the bucket.
> 
> If we now consider CloudSQL as optional (I guess users may want this too, because gcs obviously are used more often). Maybe we should just let users choose both names independently.
> 
> After all, cluster manager should remember which cloudsql and which gcs bucket were used together.

"I think we should discuss both this and #3803 together."
+1.

I just added Yuan to tomorrow meeting as optoinal. (already planned a meeting with Alexy my side tomorrow)

**Comment by rmgogogo** [comment]
"We'd want them to always use the same combination"
I also have same preference and it's the original design. Comparing with introducing the flexibility which just solve the UI level usability, keep it as same combination may better for user. 

Otherwise, if the PVC got lost, even we keep the CloudSQL, the data has no value.

**Comment by Ark-kun** [comment]
Let's discuss the issue tomorrow. But I do not understand why this PR is controversial. It's a strict improvement/bugfix. It has nothing to do with the relation between the bucket and the database.

There is a bug: #3798 . When DB connection string is not specified, the Cloud SQL is still deployed even though it just cannot work. CloudSQL cannot work when DB connection string is empty. The result is a broken deployment. This PR fixes that. No more, no less. This PR is not related to the GCS buckets at all. All this PR does is prevent using CloudSQL  when DB connection string is empty.

**Comment by Ark-kun** [comment]
>we explicitly kept sql connection name and gcs bucket name coupled
> We'd want them to always use the same combination, otherwise, they will find the data in the db doesn't match those in the bucket.

Do you think that this PR changes something in that aspect? Can you please show me the place where this PR changes that?

File: manifests/gcp_marketplace/chart/kubeflow-pipelines/templates/metadata.yaml [hunk 1]
```diff
@@ -39,7 +39,7 @@ spec:
         imagePullPolicy: 'Always'
         env:
         # TODO: merge all into mysql-credential
-        {{ if .Values.managedstorage.enabled }}
+        {{ if .Values.managedstorage.cloudsqlInstanceConnectionName }}
         - name: DBCONFIG_USER
```
Comment (rmgogogo) [inline manifests/gcp_marketplace/chart/kubeflow-pipelines/templates/metadata.yaml:43]: e.x. the case user set the field but didn't check the CheckBox (the UI give user expecation that it won't use CloudSQL).

Let's quickly review tomorrow. (I didn't aware you are still online, thanks)

**Comment by Bobgy** [comment]
@Ark-kun Sorry the discussion might have offended you. I just preferred we agree on the final CUJs first, so that we do not spend time detouring.

**Comment by Bobgy** [comment]
Thanks for your attention to user experience! Really appreciate that!

**Comment by Ark-kun** [comment]
> @Ark-kun Sorry the discussion might have offended you. I just preferred we agree on the final CUJs first, so that we do not spend time detouring.

No no no. You have not offended me in any way.
I'm also sorry. The tone of my writing might have been too harsh >_<

These PRs are not urgent for me, so we can discuss and get to the best decision.

**Comment by Bobgy** [comment]
Thanks! Let's sync in the meeting

**Comment by Ark-kun** [comment]
I think we should fix this issue. This PR prevents an invalid deployment. Can we merge it?

**Comment by Bobgy** [comment]
@rmgogogo and I prefer not putting this into 1.0 scope because it's more a marketplace UI limitation, if user read through the documentation I think there's no confusion both are required.

**Comment by Ark-kun** [comment]
> there's no confusion both are required

I found the description confused. It also confused me that clicking the "managed *storage*" checkbox resulted in a broken DB.
As we discussed before, the GCS without CloudSQL is a valid scenario. I thought we've agreed to support it.

**Comment by Bobgy** [comment]
> > there's no confusion both are required
> 
> I found the description confused. It also confused me that clicking the "managed _storage_" checkbox resulted in a broken DB.

Can you take a look at revised description from Joe in rc4? I think it became better.

> As we discussed before, the GCS without CloudSQL is a valid scenario. I thought we've agreed to support it.

I still agree with this scenario, the recent PR is more a scope down to make sure it's safe in 1.0 release. Do you want to drive the effort for supporting GCS only, I think the major efforts are documentation that can explain implications of using just GCS (how to reinstall, etc). 

My consensus with Renmin is based on marketplace having lower priority than before. Therefore we wanted to avoid the more complex scenario.

File: manifests/gcp_marketplace/chart/kubeflow-pipelines/templates/cache.yaml [hunk 2]
```diff
@@ -205,7 +205,7 @@ spec:
       - name: server
         image: {{ .Values.images.cacheserver }}
         env:
-          {{ if .Values.managedstorage.enabled }}
+          {{ if .Values.managedstorage.cloudsqlInstanceConnectionName }}
```
Comment (zijianjoy) [inline manifests/gcp_marketplace/chart/kubeflow-pipelines/templates/cache.yaml:208]: Should we check both `.Values.managedstorage.enabled ` and `cloudsqlInstanceConnectionName` for enabling cloudsql instance?

Comment (Ark-kun) [reply manifests/gcp_marketplace/chart/kubeflow-pipelines/templates/cache.yaml:208]: I wonder about this...
If the user specifies a connection, it's a bit confusing when it's not used.
Ideally we should make the cloudsqlInstanceConnectionName input enabled only when the managedstorage is set to true.

Comment (zijianjoy) [reply manifests/gcp_marketplace/chart/kubeflow-pipelines/templates/cache.yaml:208]: Do we have a mechanism to make sure `enabled` and `cloudsqlInstanceConnectionName` are exclusive to each other? So we have the confidence that user won't set both variables to be conflicting values. Sorry that I don't have much context on how marketplace manifest is provisioned, but it is good to make sure both parameters are checked, if we don't have anywhere else to enforce the input is valid.

**Comment by rimolive** [comment]
Closing this PR. No activity for more than a year.

/close

**Pull request closed without merging**